beacon_node/beacon_chain/src/execution_payload.rs

```diff
@@ -57,22 +57,26 @@ pub fn notify_new_payload<T: BeaconChainTypes>(
         .block_on(|execution_layer| execution_layer.notify_new_payload(execution_payload));
 
     match new_payload_response {
-        Ok((status, latest_valid_hash)) => match status {
-            PayloadStatusV1Status::Valid => Ok(PayloadVerificationStatus::Verified),
-            PayloadStatusV1Status::Syncing | PayloadStatusV1Status::Accepted => {
+        Ok(status) => match status {
+            PayloadStatus::Valid => Ok(PayloadVerificationStatus::Verified),
+            PayloadStatus::Syncing | PayloadStatus::Accepted => {
                 Ok(PayloadVerificationStatus::NotVerified)
             }
-            PayloadStatusV1Status::Invalid
-            | PayloadStatusV1Status::InvalidTerminalBlock
-            | PayloadStatusV1Status::InvalidBlockHash => {
-                // TODO(bellatrix): process the invalid payload.
-                //
-                // See: https://github.com/sigp/lighthouse/pull/2837
-                Err(ExecutionPayloadError::RejectedByExecutionEngine {
-                    status,
-                    latest_valid_hash,
-                }
-                .into())
+            PayloadStatus::Invalid {
+                latest_valid_hash, ..
+            } => {
+                // This block has not yet been applied to fork choice, so the latest block that was
+                // imported to fork choice was the parent.
+                let latest_root = block.parent_root();
+                chain.process_invalid_execution_payload(latest_root, Some(latest_valid_hash))?;
+
+                Err(ExecutionPayloadError::RejectedByExecutionEngine { status }.into())
+            }
+            PayloadStatus::InvalidTerminalBlock { .. } | PayloadStatus::InvalidBlockHash { .. } => {
+                // Returning an error here should sufficient to invalidate the block. We have no
+                // information to indicate it's parent is invalid, so no need to run
```

Comment:
Thinking aloud a little, we're relying on an assumption here that `notify_new_payload` is called on blocks _in order_ such that a block _must not_ have any descendants if it is found invalid at this stage.

I think that's safe, but I do wonder whether we could hit something funky in the presence of optimistic sync plus two (slashable) blocks from the same slot with different `BeaconBlockBody`s and identical `ExecutionPayload`s. The first block could be imported optimistically, with other optimistic descendants imported on top of it. Later, we import the slashable block and find that its identical payload is invalid due to the terminal block condition. We _could_ safely invalidate the descendants of the first block at this point, but will not with the code as currently written. I don't think this is a big deal however as the EE will eventually inform us of the invalidity of the descendants if we send it a fork choice update.

Also there's a typo:

```suggestion
                // information to indicate its parent is invalid, so no need to run
```

Reply:
> We could safely invalidate the descendants of the first block at this point, but will not with the code as currently written.

AIUI, the solution [in this comment](https://github.com/sigp/lighthouse/pull/2837#discussion_r815261289) has resolved this issue. Is that right?

Reply:
I was imagining that change was already implemented and trying to think of other potential issues. But I think the thing I raised is inconsequential: we might just take one extra call to the EL to work out that we should invalidate a chain that shares an invalid payload with another chain (which is only possible if we have two distinct [slashable] blocks at the same slot). It's a non-issue :relieved: